File: src/embedded/sab_queue.zig
const std = @import("std");

/// SharedArrayBuffer-based operation queue for zero N-API overhead
pub const SABQueue = struct {
    buffer: []u8,           // Points to SharedArrayBuffer memory
    head: *u32,             // Atomic head pointer (consumer)
    tail: *u32,             // Atomic tail pointer (producer)
    buffer_size: u32,
    running: std.atomic.Value(bool),
    
    // Operation types
    pub const OP_SET = 1;
    pub const OP_DELETE = 2;
    pub const OP_SHUTDOWN = 255;
    
    // Fixed-size operation header (48 bytes for alignment)
    pub const OperationHeader = packed struct {
        op_type: u32,       // Operation type
        path_offset: u32,   // Offset to path in data section
        path_length: u32,   // Length of path
        value_offset: u32,  // Offset to value in data section
        value_length: u32,  // Length of value (0 for deletes)
        sequence: u64,      // Sequence number for ordering
        reserved: u64,      // Reserved for future use
        
        comptime {
            std.debug.assert(@sizeOf(OperationHeader) == 48);
        }
    };
    
    pub fn init(sab_ptr: [*]u8, size: u32) SABQueue {
        // Layout: [head:4][tail:4][data:size-8]
        const head_ptr = @as(*u32, @ptrCast(@alignCast(sab_ptr)));
        const tail_ptr = @as(*u32, @ptrCast(@alignCast(sab_ptr + 4)));
        const data_ptr = sab_ptr + 8;
        const data_size = size - 8;
        
        // Initialize head/tail to 0
        @atomicStore(u32, head_ptr, 0, .release);
        @atomicStore(u32, tail_ptr, 0, .release);
        
        return SABQueue{
            .buffer = data_ptr[0..data_size],
            .head = head_ptr,
            .tail = tail_ptr,
            .buffer_size = data_size,
            .running = std.atomic.Value(bool).init(true),
        };
    }
    
    /// Consumer: Read next operation from queue
    pub fn dequeue(self: *SABQueue) ?OperationHeader {
        const head = @atomicLoad(u32, self.head, .acquire);
        const tail = @atomicLoad(u32, self.tail, .acquire);
        
        // Check if queue is empty
        if (head == tail) return null;
        
        // Calculate next head position
        const next_head = (head + @sizeOf(OperationHeader)) % self.buffer_size;
        
        // Ensure we don't read past tail (with wrap-around)
        const available = if (tail >= head) 
            tail - head 
        else 
            self.buffer_size - head + tail;
            
        if (available < @sizeOf(OperationHeader)) return null;
        
        // Read operation header safely (avoid alignment issues)
        const header_bytes = self.buffer[head..head + @sizeOf(OperationHeader)];
        var header: OperationHeader = undefined;
        @memcpy(std.mem.asBytes(&header), header_bytes);
        
        // Update head pointer atomically
        @atomicStore(u32, self.head, next_head, .release);
        
        return header;
    }
    
    /// Get path data for an operation
    pub fn getPath(self: *SABQueue, header: OperationHeader) []const u8 {
        const start = header.path_offset % self.buffer_size;
        const end = (start + header.path_length) % self.buffer_size;
        
        if (end > start) {
            // No wrap-around
            return self.buffer[start..end];
        } else {
            // Handle wrap-around by copying to temporary buffer
            // For now, assume no wrap-around for simplicity
            // TODO: Handle wrap-around case
            return self.buffer[start..start + header.path_length];
        }
    }
    
    /// Get value data for an operation
    pub fn getValue(self: *SABQueue, header: OperationHeader) ?[]const u8 {
        if (header.value_length == 0) return null;
        
        const start = header.value_offset % self.buffer_size;
        const end = (start + header.value_length) % self.buffer_size;
        
        if (end > start) {
            return self.buffer[start..end];
        } else {
            // Handle wrap-around case
            // TODO: Handle wrap-around properly
            return self.buffer[start..start + header.value_length];
        }
    }
    
    /// Check if we should stop processing
    pub fn shouldStop(self: *SABQueue) bool {
        return !self.running.load(.acquire);
    }
    
    /// Signal shutdown
    pub fn shutdown(self: *SABQueue) void {
        self.running.store(false, .release);
    }
    
    /// Get queue statistics
    pub fn stats(self: *SABQueue) struct { head: u32, tail: u32, pending: u32 } {
        const head = @atomicLoad(u32, self.head, .acquire);
        const tail = @atomicLoad(u32, self.tail, .acquire);
        
        const pending = if (tail >= head) 
            tail - head 
        else 
            self.buffer_size - head + tail;
            
        return .{ 
            .head = head, 
            .tail = tail, 
            .pending = pending / @sizeOf(OperationHeader) 
        };
    }
};

File: nodejs-bindings/sab-queue.js
/**
 * ArrayBuffer-based operation queue for reduced N-API overhead
 * This allows JavaScript to write operations directly to shared memory
 * that Zig can consume with minimal N-API calls!
 */

const OP_SET = 1;
const OP_DELETE = 2;
const OP_SHUTDOWN = 255;

const HEADER_SIZE = 48; // Must match Zig OperationHeader size

class SABQueue {
    constructor(arrayBuffer) {
        this.sab = arrayBuffer;
        this.view = new DataView(arrayBuffer);
        this.buffer = new Uint8Array(arrayBuffer);
        this.int32View = new Int32Array(arrayBuffer, 0, 2); // First 8 bytes for head/tail
        
        // Layout: [head:4][tail:4][data:remaining]
        this.headOffset = 0;
        this.tailOffset = 4;
        this.dataOffset = 8;
        this.dataSize = arrayBuffer.byteLength - 8;
        
        // Current data position for writing values
        this.currentDataPos = this.dataOffset;
        this.sequenceNumber = 0;
    }
    
    /**
     * Write a SET operation to the queue
     * @param {string} path - Database path
     * @param {any} value - Value to store
     * @returns {boolean} Success
     */
    enqueueSet(path, value) {
        const pathBytes = new TextEncoder().encode(path);
        const valueBytes = this.encodeValue(value);
        
        return this.enqueueOperation(OP_SET, pathBytes, valueBytes);
    }
    
    /**
     * Write a DELETE operation to the queue
     * @param {string} path - Database path
     * @returns {boolean} Success
     */
    enqueueDelete(path) {
        const pathBytes = new TextEncoder().encode(path);
        
        return this.enqueueOperation(OP_DELETE, pathBytes, null);
    }
    
    /**
     * Encode JavaScript value to binary format that Zig can understand
     */
    encodeValue(value) {
        if (typeof value === 'string') {
            // 's' + string bytes
            const strBytes = new TextEncoder().encode(value);
            const result = new Uint8Array(1 + strBytes.length);
            result[0] = 115; // 's'
            result.set(strBytes, 1);
            return result;
        }
        
        if (typeof value === 'number') {
            // 'n' + 8 bytes for f64
            const result = new Uint8Array(9);
            result[0] = 110; // 'n'
            const view = new DataView(result.buffer);
            view.setFloat64(1, value, true); // little endian
            return result;
        }
        
        if (typeof value === 'boolean') {
            // 'b' + 1 byte
            const result = new Uint8Array(2);
            result[0] = 98; // 'b'
            result[1] = value ? 1 : 0;
            return result;
        }
        
        if (value === null || value === undefined) {
            // 'z' for null
            return new Uint8Array([122]); // 'z'
        }
        
        // For complex objects, encode as JSON
        // 'j' + JSON string bytes
        const jsonStr = JSON.stringify(value);
        const jsonBytes = new TextEncoder().encode(jsonStr);
        const result = new Uint8Array(1 + jsonBytes.length);
        result[0] = 106; // 'j'
        result.set(jsonBytes, 1);
        return result;
    }
    
    /**
     * Low-level operation enqueue
     */
    enqueueOperation(opType, pathBytes, valueBytes) {
        const valueLength = valueBytes ? valueBytes.length : 0;
        const totalSize = HEADER_SIZE + pathBytes.length + valueLength;
        
        // Check if we have space (simple check, doesn't handle wrap-around perfectly)
        const currentTail = this.int32View[this.tailOffset / 4];
        const currentHead = this.int32View[this.headOffset / 4];
        
        const available = this.dataSize - (currentTail - currentHead);
        if (totalSize > available) {
            console.warn('SAB queue full, dropping operation');
            return false;
        }
        
        // Calculate positions
        const headerPos = currentTail;
        const pathPos = this.allocateData(pathBytes.length);
        const valuePos = valueBytes ? this.allocateData(valueLength) : 0;
        
        // Write path data
        this.buffer.set(pathBytes, pathPos);
        
        // Write value data
        if (valueBytes) {
            this.buffer.set(valueBytes, valuePos);
        }
        
        // Write operation header
        this.writeHeader(headerPos, opType, pathPos, pathBytes.length, valuePos, valueLength);
        
        // Update tail pointer
        const newTail = (currentTail + HEADER_SIZE) % this.dataSize;
        this.int32View[this.tailOffset / 4] = newTail;
        
        return true;
    }
    
    /**
     * Allocate space in data section
     */
    allocateData(size) {
        const pos = this.currentDataPos;
        this.currentDataPos += size;
        
        // Simple wrap-around (not perfect, but good enough for demo)
        if (this.currentDataPos >= this.dataOffset + this.dataSize) {
            this.currentDataPos = this.dataOffset + 1024; // Leave some space after headers
        }
        
        return pos;
    }
    
    /**
     * Write operation header to specific position
     */
    writeHeader(pos, opType, pathOffset, pathLength, valueOffset, valueLength) {
        const headerView = new DataView(this.sab, pos, HEADER_SIZE);
        
        headerView.setUint32(0, opType, true);           // op_type (4 bytes)
        headerView.setUint32(4, pathOffset, true);       // path_offset (4 bytes)
        headerView.setUint32(8, pathLength, true);       // path_length (4 bytes)
        headerView.setUint32(12, valueOffset, true);     // value_offset (4 bytes)
        headerView.setUint32(16, valueLength, true);     // value_length (4 bytes)
        headerView.setBigUint64(20, BigInt(this.sequenceNumber++), true); // sequence (8 bytes)
        headerView.setBigUint64(28, 0n, true);           // reserved (8 bytes)
        // Total: 4+4+4+4+4+8+8 = 36 bytes, we need 12 more bytes for 48 total
        headerView.setUint32(36, 0, true);               // padding
        headerView.setUint32(40, 0, true);               // padding
        headerView.setUint32(44, 0, true);               // padding (total now 48 bytes)
    }
    
    /**
     * Get queue statistics
     */
    getStats() {
        const head = this.int32View[this.headOffset / 4];
        const tail = this.int32View[this.tailOffset / 4];
        
        const pending = tail >= head ? tail - head : this.dataSize - head + tail;
        
        return {
            head,
            tail,
            pending: Math.floor(pending / HEADER_SIZE),
            bufferSize: this.dataSize
        };
    }
    
    /**
     * Signal shutdown to worker thread
     */
    shutdown() {
        return this.enqueueOperation(OP_SHUTDOWN, new Uint8Array([]), null);
    }
}

module.exports = { SABQueue, OP_SET, OP_DELETE, OP_SHUTDOWN };